main/tests/server/src/com/google/refine/grel/controls/FilterTests.java

```diff
@@ -27,15 +27,53 @@
 
 package com.google.refine.grel.controls;
 
+import static org.testng.Assert.assertTrue;
+import static org.testng.Assert.fail;
+
+import java.util.Properties;
+
 import org.testng.annotations.Test;
 
+import com.google.refine.expr.EvalError;
+import com.google.refine.expr.Evaluable;
+import com.google.refine.expr.MetaParser;
+import com.google.refine.expr.ParsingException;
+import com.google.refine.grel.GrelTestBase;
 import com.google.refine.util.TestUtils;
 
-public class FilterTests {
+public class FilterTests extends GrelTestBase {
 
     @Test
     public void serializeFilter() {
         String json = "{\"description\":\"Evaluates expression a to an array. Then for each array element, binds its value to variable name v, evaluates expression test which should return a boolean. If the boolean is true, pushes v onto the result array.\",\"params\":\"expression a, variable v, expression test\",\"returns\":\"array\"}";
         TestUtils.isSerializedTo(new Filter(), json);
     }
+
+    private void assertParseError(String expression) throws ParsingException {
+        Evaluable eval = MetaParser.parse("grel:" + expression);
+        Object result = eval.evaluate(bindings);
+        assertTrue(result instanceof EvalError, "Expression didn't return error : " + expression);
+    }
+
+    @Test
+    public void testInvalidParams() throws ParsingException {
+        bindings = new Properties();
+        bindings.put("v", "");
+        assertParseError("filter('test', v, v)");
+        try {
+            assertParseError("filter([], 1, 1)");
+            fail("Didn't throw a ParsingException for wrong argument type");
+        } catch (ParsingException e) {
+        }
```

Comment:
As mentioned in my earlier comment, it is better if you can use `assertThrows` to assert that a particular piece of code throws an exception. Here is an example to help you understand what it would look like:
```suggestion
        assertThrows("Didn't throw a ParsingException for wrong argument type", ParsingException.class,
                () -> MetaParser.parse("grel:filter([], 1, 1)"));
```
Because the error is coming from the parsing stage, not the evaluation stage, we can make the assertion narrower by calling the parser only and checking that it does throw.

Would you be able to adapt the rest of the test following this pattern?

Reply:
@wetneb I updated all the cases to use assertThrows